{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Create Sequences for Dataset"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "import copy\n",
    "import gumpy"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Get sequences from mutations"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "reference = gumpy.Genome('../data/NC_000962.3.gbk')\n",
    "pnca = reference.build_gene('pncA')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# from fowler-lab/predict-pyrazinamide-resistance GitHub\n",
    "\n",
    "pnca_paper_dataset = pd.read_csv('../data/ds-traintest-phen.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>MUTATION</th>\n",
       "      <th>CONSISTENT_PHENOTYPE</th>\n",
       "      <th>REF</th>\n",
       "      <th>AMINO_ACID</th>\n",
       "      <th>ALT</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>A102V</td>\n",
       "      <td>S</td>\n",
       "      <td>A</td>\n",
       "      <td>102</td>\n",
       "      <td>V</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>A134D</td>\n",
       "      <td>S</td>\n",
       "      <td>A</td>\n",
       "      <td>134</td>\n",
       "      <td>D</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>A134P</td>\n",
       "      <td>R</td>\n",
       "      <td>A</td>\n",
       "      <td>134</td>\n",
       "      <td>P</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>A134S</td>\n",
       "      <td>S</td>\n",
       "      <td>A</td>\n",
       "      <td>134</td>\n",
       "      <td>S</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>A134V</td>\n",
       "      <td>R</td>\n",
       "      <td>A</td>\n",
       "      <td>134</td>\n",
       "      <td>V</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>A143D</td>\n",
       "      <td>R</td>\n",
       "      <td>A</td>\n",
       "      <td>143</td>\n",
       "      <td>D</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>A143G</td>\n",
       "      <td>R</td>\n",
       "      <td>A</td>\n",
       "      <td>143</td>\n",
       "      <td>G</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>A143V</td>\n",
       "      <td>S</td>\n",
       "      <td>A</td>\n",
       "      <td>143</td>\n",
       "      <td>V</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>A146E</td>\n",
       "      <td>R</td>\n",
       "      <td>A</td>\n",
       "      <td>146</td>\n",
       "      <td>E</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>A146S</td>\n",
       "      <td>R</td>\n",
       "      <td>A</td>\n",
       "      <td>146</td>\n",
       "      <td>S</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  MUTATION CONSISTENT_PHENOTYPE REF  AMINO_ACID ALT\n",
       "0    A102V                    S   A         102   V\n",
       "1    A134D                    S   A         134   D\n",
       "2    A134P                    R   A         134   P\n",
       "3    A134S                    S   A         134   S\n",
       "4    A134V                    R   A         134   V\n",
       "5    A143D                    R   A         143   D\n",
       "6    A143G                    R   A         143   G\n",
       "7    A143V                    S   A         143   V\n",
       "8    A146E                    R   A         146   E\n",
       "9    A146S                    R   A         146   S"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "def split_mutation(row):\n",
    "    return pd.Series([row.MUTATION[0], int(row.MUTATION[1:-1]), row.MUTATION[-1]])\n",
    "\n",
    "pnca_paper_dataset[['REF','AMINO_ACID','ALT']] = pnca_paper_dataset.apply(split_mutation,axis=1)\n",
    "\n",
    "pnca_paper_dataset.head(10)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "# create table of sample sequences\n",
    "\n",
    "sequences = pd.DataFrame(columns=['phenotype_label','number_resistant_mutations', 'number_susceptible_mutations', 'allele', 'mutation'])\n",
    "\n",
    "for idx, row in pnca_paper_dataset.iterrows():\n",
    "    \n",
    "    sample = copy.deepcopy(pnca)\n",
    "    # reassign amino acid to mutated amino acid\n",
    "    sample.amino_acid_sequence[sample.amino_acid_number==row.AMINO_ACID] = row.ALT\n",
    "    \n",
    "    sequence = ''.join(i for i in sample.amino_acid_sequence)\n",
    "    mutation = row.MUTATION\n",
    "    \n",
    "    if row.CONSISTENT_PHENOTYPE == 'R':\n",
    "        number_resistant_mutations = 1\n",
    "        number_susceptible_mutations = 0\n",
    "        phenotype_label = 'R'\n",
    "    else:\n",
    "        number_resistant_mutations = 0\n",
    "        number_susceptible_mutations = 1\n",
    "        phenotype_label = 'S'\n",
    "        \n",
    "    sequences.loc[len(sequences)] = [phenotype_label, number_resistant_mutations, number_susceptible_mutations, sequence, mutation]\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>phenotype_label</th>\n",
       "      <th>number_resistant_mutations</th>\n",
       "      <th>number_susceptible_mutations</th>\n",
       "      <th>allele</th>\n",
       "      <th>mutation</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>S</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>MRALIIVDVQNDFCEGGSLAVTGGAALARAISDYLAEAADYHHVVA...</td>\n",
       "      <td>A102V</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>S</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>MRALIIVDVQNDFCEGGSLAVTGGAALARAISDYLAEAADYHHVVA...</td>\n",
       "      <td>A134D</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>R</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>MRALIIVDVQNDFCEGGSLAVTGGAALARAISDYLAEAADYHHVVA...</td>\n",
       "      <td>A134P</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>S</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>MRALIIVDVQNDFCEGGSLAVTGGAALARAISDYLAEAADYHHVVA...</td>\n",
       "      <td>A134S</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>R</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>MRALIIVDVQNDFCEGGSLAVTGGAALARAISDYLAEAADYHHVVA...</td>\n",
       "      <td>A134V</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>659</th>\n",
       "      <td>S</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>MRALIIVDVQNDFCEGGSLAVTGGAALARAISDYLAEAADYHHVVA...</td>\n",
       "      <td>Y95N</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>660</th>\n",
       "      <td>S</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>MRALIIVDVQNDFCEGGSLAVTGGAALARAISDYLAEAADYHHVVA...</td>\n",
       "      <td>Y99C</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>661</th>\n",
       "      <td>S</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>MRALIIVDVQNDFCEGGSLAVTGGAALARAISDYLAEAADYHHVVA...</td>\n",
       "      <td>Y99D</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>662</th>\n",
       "      <td>S</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>MRALIIVDVQNDFCEGGSLAVTGGAALARAISDYLAEAADYHHVVA...</td>\n",
       "      <td>Y99F</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>663</th>\n",
       "      <td>S</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>MRALIIVDVQNDFCEGGSLAVTGGAALARAISDYLAEAADYHHVVA...</td>\n",
       "      <td>Y99N</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>664 rows × 5 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "    phenotype_label  number_resistant_mutations  number_susceptible_mutations  \\\n",
       "0                 S                           0                             1   \n",
       "1                 S                           0                             1   \n",
       "2                 R                           1                             0   \n",
       "3                 S                           0                             1   \n",
       "4                 R                           1                             0   \n",
       "..              ...                         ...                           ...   \n",
       "659               S                           0                             1   \n",
       "660               S                           0                             1   \n",
       "661               S                           0                             1   \n",
       "662               S                           0                             1   \n",
       "663               S                           0                             1   \n",
       "\n",
       "                                                allele mutation  \n",
       "0    MRALIIVDVQNDFCEGGSLAVTGGAALARAISDYLAEAADYHHVVA...    A102V  \n",
       "1    MRALIIVDVQNDFCEGGSLAVTGGAALARAISDYLAEAADYHHVVA...    A134D  \n",
       "2    MRALIIVDVQNDFCEGGSLAVTGGAALARAISDYLAEAADYHHVVA...    A134P  \n",
       "3    MRALIIVDVQNDFCEGGSLAVTGGAALARAISDYLAEAADYHHVVA...    A134S  \n",
       "4    MRALIIVDVQNDFCEGGSLAVTGGAALARAISDYLAEAADYHHVVA...    A134V  \n",
       "..                                                 ...      ...  \n",
       "659  MRALIIVDVQNDFCEGGSLAVTGGAALARAISDYLAEAADYHHVVA...     Y95N  \n",
       "660  MRALIIVDVQNDFCEGGSLAVTGGAALARAISDYLAEAADYHHVVA...     Y99C  \n",
       "661  MRALIIVDVQNDFCEGGSLAVTGGAALARAISDYLAEAADYHHVVA...     Y99D  \n",
       "662  MRALIIVDVQNDFCEGGSLAVTGGAALARAISDYLAEAADYHHVVA...     Y99F  \n",
       "663  MRALIIVDVQNDFCEGGSLAVTGGAALARAISDYLAEAADYHHVVA...     Y99N  \n",
       "\n",
       "[664 rows x 5 columns]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/plain": [
       "phenotype_label\n",
       "R    349\n",
       "S    315\n",
       "Name: count, dtype: int64"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "display(sequences)\n",
    "display(sequences.phenotype_label.value_counts())"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Get original Train / Test Split"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# from fowler-lab/predict-pyrazinamide-resistance GitHub\n",
    "\n",
    "with open('../data/ds-train.npy', 'rb') as f:\n",
    "    y_train = np.load(f)\n",
    "    x_train = np.load(f)\n",
    "    train_muts = np.load(f, allow_pickle=True)\n",
    "    \n",
    "with open('../data/ds-test.npy', 'rb') as f:\n",
    "    y_test = np.load(f)\n",
    "    x_test = np.load(f)\n",
    "    test_muts = np.load(f, allow_pickle=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>phenotype_label</th>\n",
       "      <th>number_resistant_mutations</th>\n",
       "      <th>number_susceptible_mutations</th>\n",
       "      <th>allele</th>\n",
       "      <th>mutation</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>S</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>MRALIIVDVQNDFCEGGSLAVTGGAALARAISDYLAEAADYHHVVA...</td>\n",
       "      <td>A102V</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>S</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>MRALIIVDVQNDFCEGGSLAVTGGAALARAISDYLAEAADYHHVVA...</td>\n",
       "      <td>A134D</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>R</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>MRALIIVDVQNDFCEGGSLAVTGGAALARAISDYLAEAADYHHVVA...</td>\n",
       "      <td>A134P</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>S</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>MRALIIVDVQNDFCEGGSLAVTGGAALARAISDYLAEAADYHHVVA...</td>\n",
       "      <td>A134S</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>R</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>MRALIIVDVQNDFCEGGSLAVTGGAALARAISDYLAEAADYHHVVA...</td>\n",
       "      <td>A134V</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>459</th>\n",
       "      <td>S</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>MRALIIVDVQNDFCEGGSLAVTGGAALARAISDHLAEAADYHHVVA...</td>\n",
       "      <td>Y34H</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>460</th>\n",
       "      <td>S</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>MRALIIVDVQNDFCEGGSLAVTGGAALARAISDYLAEAADYHHVVA...</td>\n",
       "      <td>Y64N</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>461</th>\n",
       "      <td>S</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>MRALIIVDVQNDFCEGGSLAVTGGAALARAISDYLAEAADYHHVVA...</td>\n",
       "      <td>Y95H</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>462</th>\n",
       "      <td>S</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>MRALIIVDVQNDFCEGGSLAVTGGAALARAISDYLAEAADYHHVVA...</td>\n",
       "      <td>Y99C</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>463</th>\n",
       "      <td>S</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>MRALIIVDVQNDFCEGGSLAVTGGAALARAISDYLAEAADYHHVVA...</td>\n",
       "      <td>Y99N</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>464 rows × 5 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "    phenotype_label  number_resistant_mutations  number_susceptible_mutations  \\\n",
       "0                 S                           0                             1   \n",
       "1                 S                           0                             1   \n",
       "2                 R                           1                             0   \n",
       "3                 S                           0                             1   \n",
       "4                 R                           1                             0   \n",
       "..              ...                         ...                           ...   \n",
       "459               S                           0                             1   \n",
       "460               S                           0                             1   \n",
       "461               S                           0                             1   \n",
       "462               S                           0                             1   \n",
       "463               S                           0                             1   \n",
       "\n",
       "                                                allele mutation  \n",
       "0    MRALIIVDVQNDFCEGGSLAVTGGAALARAISDYLAEAADYHHVVA...    A102V  \n",
       "1    MRALIIVDVQNDFCEGGSLAVTGGAALARAISDYLAEAADYHHVVA...    A134D  \n",
       "2    MRALIIVDVQNDFCEGGSLAVTGGAALARAISDYLAEAADYHHVVA...    A134P  \n",
       "3    MRALIIVDVQNDFCEGGSLAVTGGAALARAISDYLAEAADYHHVVA...    A134S  \n",
       "4    MRALIIVDVQNDFCEGGSLAVTGGAALARAISDYLAEAADYHHVVA...    A134V  \n",
       "..                                                 ...      ...  \n",
       "459  MRALIIVDVQNDFCEGGSLAVTGGAALARAISDHLAEAADYHHVVA...     Y34H  \n",
       "460  MRALIIVDVQNDFCEGGSLAVTGGAALARAISDYLAEAADYHHVVA...     Y64N  \n",
       "461  MRALIIVDVQNDFCEGGSLAVTGGAALARAISDYLAEAADYHHVVA...     Y95H  \n",
       "462  MRALIIVDVQNDFCEGGSLAVTGGAALARAISDYLAEAADYHHVVA...     Y99C  \n",
       "463  MRALIIVDVQNDFCEGGSLAVTGGAALARAISDYLAEAADYHHVVA...     Y99N  \n",
       "\n",
       "[464 rows x 5 columns]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>phenotype_label</th>\n",
       "      <th>number_resistant_mutations</th>\n",
       "      <th>number_susceptible_mutations</th>\n",
       "      <th>allele</th>\n",
       "      <th>mutation</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>R</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>MRALIIVDVQNDFCEGGSLAVTGGAALARAISDYLAEAADYHHVVA...</td>\n",
       "      <td>A143D</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>S</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>MRALIIVDVQNDFCEGGSLAVTGGAALARAISDYLAEAADYHHVVA...</td>\n",
       "      <td>A143V</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>R</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>MRALIIVDVQNDFCEGGSLAVTGGAALARAISDYLAEAADYHHVVA...</td>\n",
       "      <td>A146E</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>S</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>MRALIIVDVQNDFCEGGSLAVTGGAALARAISDYLAEAADYHHVVA...</td>\n",
       "      <td>A161T</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>S</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>MRALIIVDVQNDFCEGGSLAVTGGAALARAISDYLAEAADYHHVVA...</td>\n",
       "      <td>A178D</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>195</th>\n",
       "      <td>R</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>MRALIIVDVQNDFCEGGSLAVTGGAALARAISDYLAEAADHHHVVA...</td>\n",
       "      <td>Y41H</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>196</th>\n",
       "      <td>S</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>MRALIIVDVQNDFCEGGSLAVTGGAALARAISDYLAEAADYHHVVA...</td>\n",
       "      <td>Y95C</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>197</th>\n",
       "      <td>S</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>MRALIIVDVQNDFCEGGSLAVTGGAALARAISDYLAEAADYHHVVA...</td>\n",
       "      <td>Y95N</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>198</th>\n",
       "      <td>S</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>MRALIIVDVQNDFCEGGSLAVTGGAALARAISDYLAEAADYHHVVA...</td>\n",
       "      <td>Y99D</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>199</th>\n",
       "      <td>S</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>MRALIIVDVQNDFCEGGSLAVTGGAALARAISDYLAEAADYHHVVA...</td>\n",
       "      <td>Y99F</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>200 rows × 5 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "    phenotype_label  number_resistant_mutations  number_susceptible_mutations  \\\n",
       "0                 R                           1                             0   \n",
       "1                 S                           0                             1   \n",
       "2                 R                           1                             0   \n",
       "3                 S                           0                             1   \n",
       "4                 S                           0                             1   \n",
       "..              ...                         ...                           ...   \n",
       "195               R                           1                             0   \n",
       "196               S                           0                             1   \n",
       "197               S                           0                             1   \n",
       "198               S                           0                             1   \n",
       "199               S                           0                             1   \n",
       "\n",
       "                                                allele mutation  \n",
       "0    MRALIIVDVQNDFCEGGSLAVTGGAALARAISDYLAEAADYHHVVA...    A143D  \n",
       "1    MRALIIVDVQNDFCEGGSLAVTGGAALARAISDYLAEAADYHHVVA...    A143V  \n",
       "2    MRALIIVDVQNDFCEGGSLAVTGGAALARAISDYLAEAADYHHVVA...    A146E  \n",
       "3    MRALIIVDVQNDFCEGGSLAVTGGAALARAISDYLAEAADYHHVVA...    A161T  \n",
       "4    MRALIIVDVQNDFCEGGSLAVTGGAALARAISDYLAEAADYHHVVA...    A178D  \n",
       "..                                                 ...      ...  \n",
       "195  MRALIIVDVQNDFCEGGSLAVTGGAALARAISDYLAEAADHHHVVA...     Y41H  \n",
       "196  MRALIIVDVQNDFCEGGSLAVTGGAALARAISDYLAEAADYHHVVA...     Y95C  \n",
       "197  MRALIIVDVQNDFCEGGSLAVTGGAALARAISDYLAEAADYHHVVA...     Y95N  \n",
       "198  MRALIIVDVQNDFCEGGSLAVTGGAALARAISDYLAEAADYHHVVA...     Y99D  \n",
       "199  MRALIIVDVQNDFCEGGSLAVTGGAALARAISDYLAEAADYHHVVA...     Y99F  \n",
       "\n",
       "[200 rows x 5 columns]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "train_muts_df = pd.DataFrame({'mutation': list(train_muts)})\n",
    "test_muts_df = pd.DataFrame({'mutation': list(test_muts)})\n",
    "\n",
    "train_seqs = pd.merge(sequences, train_muts_df, on='mutation', how='inner')\n",
    "test_seqs = pd.merge(sequences, test_muts_df, on='mutation', how='inner')   \n",
    "\n",
    "display(train_seqs)\n",
    "display(test_seqs)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "with open('../data/real_train_sequences.csv', 'wb') as file:\n",
    "    train_seqs.to_csv(file, index=False)\n",
    "\n",
    "with open('../data/real_test_sequences.csv', 'wb') as file:\n",
    "    test_seqs.to_csv(file, index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Create FASTA files for AlphaFold"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "def save_fasta(header, sequence, name):\n",
    "    filename = f\"../data/fastas/{name}.fasta\"\n",
    "    \"\"\"Save a sequence in FASTA format to a file.\"\"\"\n",
    "    with open(filename, \"w\") as fasta_file:\n",
    "        fasta_file.write(f\">{header}\\n\")\n",
    "        fasta_file.write(sequence)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [],
   "source": [
    "filename = \"../data/fastas/train.fasta\"\n",
    "\n",
    "with open(filename, 'w') as fasta_file:\n",
    "    \n",
    "    for index, seq_row in train_seqs.iterrows():\n",
    "        res_index = int(seq_row.mutation[1:-1]) -1\n",
    "        alt_res = seq_row.mutation[-1]\n",
    "        assert alt_res == seq_row.allele[res_index], 'Mutation does not match sequence \\n' + str(seq_row.mutation) + '\\n' + str(seq_row.allele)\n",
    "\n",
    "        sequence = seq_row.allele\n",
    "        \n",
    "        fasta_file.write(f\">{index}\\n\")\n",
    "        fasta_file.write(f\"{sequence}\\n\")\n",
    "        "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [],
   "source": [
    "filename = \"../data/fastas/test.fasta\"\n",
    "\n",
    "with open(filename, 'w') as fasta_file:\n",
    "    \n",
    "    for index, seq_row in test_seqs.iterrows():\n",
    "        res_index = int(seq_row.mutation[1:-1]) -1\n",
    "        alt_res = seq_row.mutation[-1]\n",
    "        assert alt_res == seq_row.allele[res_index], 'Mutation does not match sequence \\n' + str(seq_row.mutation) + '\\n' + str(seq_row.allele)\n",
    "\n",
    "        sequence = seq_row.allele\n",
    "        \n",
    "        fasta_file.write(f\">{index}\\n\")\n",
    "        fasta_file.write(f\"{sequence}\\n\")"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.0"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
