{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import pickle as pkl\n",
    "import tqdm"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [],
   "source": [
    "train_seqs = pd.read_csv('../data/real_train_sequences.csv')\n",
    "test_seqs = pd.read_csv('../data/real_test_sequences.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "with open('../data/original-ds-muts.pkl', 'rb') as file:\n",
    "    mutations = pkl.load(file)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "mutations['train'].sort()\n",
    "mutations['test'].sort()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['A143D',\n",
       " 'A143V',\n",
       " 'A146E',\n",
       " 'A161T',\n",
       " 'A178D',\n",
       " 'A20E',\n",
       " 'A20S',\n",
       " 'A20T',\n",
       " 'A28P',\n",
       " 'A30D',\n",
       " 'A30P',\n",
       " 'A30S',\n",
       " 'A36S',\n",
       " 'A36V',\n",
       " 'A38S',\n",
       " 'A39S',\n",
       " 'A3V',\n",
       " 'A46E',\n",
       " 'A46S',\n",
       " 'A79S',\n",
       " 'A79T',\n",
       " 'A89P',\n",
       " 'A92S',\n",
       " 'C138S',\n",
       " 'C184Y',\n",
       " 'D110N',\n",
       " 'D110V',\n",
       " 'D129E',\n",
       " 'D12E',\n",
       " 'D158N',\n",
       " 'D166V',\n",
       " 'D49E',\n",
       " 'D49G',\n",
       " 'D49Y',\n",
       " 'D56N',\n",
       " 'D63N',\n",
       " 'D80N',\n",
       " 'D86E',\n",
       " 'D86N',\n",
       " 'D8E',\n",
       " 'D8G',\n",
       " 'D8H',\n",
       " 'E107Q',\n",
       " 'E127D',\n",
       " 'E127K',\n",
       " 'E144G',\n",
       " 'E15D',\n",
       " 'E15G',\n",
       " 'E15Q',\n",
       " 'E173K',\n",
       " 'E173V',\n",
       " 'E174G',\n",
       " 'E181K',\n",
       " 'E91D',\n",
       " 'F13I',\n",
       " 'F13V',\n",
       " 'F13Y',\n",
       " 'F50Y',\n",
       " 'F58I',\n",
       " 'F58L',\n",
       " 'F58Y',\n",
       " 'F81S',\n",
       " 'F94C',\n",
       " 'G105S',\n",
       " 'G108R',\n",
       " 'G124A',\n",
       " 'G124C',\n",
       " 'G124D',\n",
       " 'G132R',\n",
       " 'G16S',\n",
       " 'G16V',\n",
       " 'G17S',\n",
       " 'G17V',\n",
       " 'G23A',\n",
       " 'G23V',\n",
       " 'G24D',\n",
       " 'G60D',\n",
       " 'G75D',\n",
       " 'G75V',\n",
       " 'G78D',\n",
       " 'G97D',\n",
       " 'G97V',\n",
       " 'H137Q',\n",
       " 'H43P',\n",
       " 'H43Y',\n",
       " 'H57P',\n",
       " 'H57Q',\n",
       " 'H57R',\n",
       " 'H71D',\n",
       " 'H71N',\n",
       " 'H71P',\n",
       " 'H82L',\n",
       " 'H82R',\n",
       " 'I133N',\n",
       " 'I31N',\n",
       " 'I52N',\n",
       " 'I6F',\n",
       " 'I90F',\n",
       " 'I90S',\n",
       " 'I90T',\n",
       " 'K48T',\n",
       " 'K96Q',\n",
       " 'K96T',\n",
       " 'L116P',\n",
       " 'L116R',\n",
       " 'L120R',\n",
       " 'L156P',\n",
       " 'L159Q',\n",
       " 'L172P',\n",
       " 'L172Q',\n",
       " 'L172R',\n",
       " 'L182M',\n",
       " 'L182W',\n",
       " 'L19V',\n",
       " 'L35Q',\n",
       " 'L35R',\n",
       " 'L35V',\n",
       " 'L4F',\n",
       " 'L4V',\n",
       " 'L4W',\n",
       " 'M175I',\n",
       " 'M175K',\n",
       " 'M175L',\n",
       " 'M1V',\n",
       " 'N112S',\n",
       " 'N112Y',\n",
       " 'N118Y',\n",
       " 'N11H',\n",
       " 'N11I',\n",
       " 'N149D',\n",
       " 'N149H',\n",
       " 'N149I',\n",
       " 'N149K',\n",
       " 'P115A',\n",
       " 'P54T',\n",
       " 'P62L',\n",
       " 'Q10H',\n",
       " 'Q10P',\n",
       " 'Q10R',\n",
       " 'Q122L',\n",
       " 'Q141E',\n",
       " 'R123C',\n",
       " 'R123G',\n",
       " 'R123H',\n",
       " 'R123L',\n",
       " 'R140P',\n",
       " 'R148L',\n",
       " 'R176G',\n",
       " 'R176S',\n",
       " 'R29C',\n",
       " 'S104I',\n",
       " 'S185G',\n",
       " 'S185I',\n",
       " 'S18P',\n",
       " 'S18T',\n",
       " 'S32I',\n",
       " 'S32R',\n",
       " 'S59T',\n",
       " 'S67P',\n",
       " 'S84R',\n",
       " 'S88P',\n",
       " 'S88T',\n",
       " 'T100A',\n",
       " 'T135P',\n",
       " 'T142R',\n",
       " 'T160I',\n",
       " 'T168A',\n",
       " 'T177I',\n",
       " 'T22A',\n",
       " 'T76S',\n",
       " 'V128G',\n",
       " 'V130E',\n",
       " 'V131D',\n",
       " 'V139E',\n",
       " 'V139L',\n",
       " 'V157A',\n",
       " 'V163E',\n",
       " 'V163G',\n",
       " 'V163M',\n",
       " 'V169F',\n",
       " 'V180G',\n",
       " 'V183I',\n",
       " 'V21G',\n",
       " 'V21L',\n",
       " 'V45A',\n",
       " 'V45L',\n",
       " 'V73L',\n",
       " 'V7D',\n",
       " 'V7F',\n",
       " 'V93L',\n",
       " 'W119G',\n",
       " 'W119R',\n",
       " 'W68G',\n",
       " 'W68L',\n",
       " 'Y103D',\n",
       " 'Y41H',\n",
       " 'Y95C',\n",
       " 'Y95N',\n",
       " 'Y99D',\n",
       " 'Y99F']"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "mutations['test']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'MRALIIVDVQNDFCEGGSLAVTGGAALARAISDYLAEAADYHHVVATKDFHIDPGDHFSGTPDYSSSWPPHCVSGTPGADFHPSLDTSAIEAVFYKGAYTGVYSGFEGVDENGTPLLNWLRQRGVDEVDVVGIATDHCVRQTAEDAVRNGLATRVLVDLTAGVSADTTVAALEEMRTASVELVCSS!'"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "(train_seqs.allele.values[0])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "# train_seqs['mutation'] = mutations['train']\n",
    "# test_seqs['mutation'] = mutations['test']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [],
   "source": [
    "with open('../data/real_train_sequences.csv', 'wb') as file:\n",
    "    train_seqs.to_csv(file, index=False)\n",
    "\n",
    "with open('../data/real_test_sequences.csv', 'wb') as file:\n",
    "    test_seqs.to_csv(file, index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "def save_fasta(header, sequence, name):\n",
    "    filename = f\"../data/fastas/{name}.fasta\"\n",
    "    \"\"\"Save a sequence in FASTA format to a file.\"\"\"\n",
    "    with open(filename, \"w\") as fasta_file:\n",
    "        fasta_file.write(f\">{header}\\n\")\n",
    "        fasta_file.write(sequence)\n",
    "\n",
    "# Example usage\n",
    "header = \"Example_Sequence\"\n",
    "sequence = \"AGCTGATCGTAGCTAGCTGATCGTAGCTGATGCTAGCTGATGCTAGCTGATCGT\" * 2  # Example sequence\n",
    "save_fasta(header, sequence, \"test\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "filename = \"../data/fastas/train.fasta\"\n",
    "\n",
    "with open(filename, 'w') as fasta_file:\n",
    "    \n",
    "    for mut, (index, seq_row) in zip(mutations['train'], train_seqs.iterrows()):\n",
    "        res_index = int(mut[1:-1]) -1\n",
    "        alt_res = mut[-1]\n",
    "        assert alt_res == seq_row.allele[res_index], 'Mutation does not match sequence \\n' + str(mut) + '\\n' + str(seq_row.allele)\n",
    "\n",
    "        sequence = seq_row.allele\n",
    "        \n",
    "        fasta_file.write(f\">{index}\\n\")\n",
    "        fasta_file.write(f\"{sequence}\\n\")\n",
    "        \n",
    "        # print(mut)\n",
    "        # print(seq_row.allele)\n",
    "        \n",
    "        # save_fasta(f'3PL1 {mut} variant|Chain A|PNCA ', seq_row.allele, f'train/{mut}')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [],
   "source": [
    "filename = \"../data/fastas/test.fasta\"\n",
    "\n",
    "with open(filename, 'w') as fasta_file:\n",
    "    \n",
    "    for mut, (index, seq_row) in zip(mutations['test'], test_seqs.iterrows()):\n",
    "        res_index = int(mut[1:-1]) -1\n",
    "        alt_res = mut[-1]\n",
    "        assert alt_res == seq_row.allele[res_index], 'Mutation does not match sequence \\n' + str(mut) + '\\n' + str(seq_row.allele)\n",
    "\n",
    "        sequence = seq_row.allele\n",
    "        \n",
    "        fasta_file.write(f\">{index}\\n\")\n",
    "        fasta_file.write(f\"{sequence}\\n\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [],
   "source": [
    "# for mut, (_, seq_row) in zip(mutations['test'], test_seqs.iterrows()):\n",
    "#     res_index = int(mut[1:-1]) -1\n",
    "#     alt_res = mut[-1]\n",
    "#     assert alt_res == seq_row.allele[res_index], 'Mutation does not match sequence \\n' + str(mut) + '\\n' + str(seq_row.allele)\n",
    "\n",
    "#     save_fasta(f'3PL1 {mut} variant|Chain A|PNCA ', seq_row.allele, f'test/{mut}')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.0"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
